## Wstępna analiza

Wczytujemy plik csv używając biblioteki pandas jako DataFrame, upewniamy się że widoczne będą wszystkie kolumny z danymi, analizujemy wielkość i naturę zbioru danych.

In [19]:
import pandas as pd
df = pd.read_csv('data/final_train.csv')
pd.set_option('display.max_columns', None)

In [20]:
df.shape

(11086, 573)

In [21]:
df.sample(6)

Unnamed: 0          Activity  angle(X,gravityMean)  \
5954        5954           WALKING             -1.574491   
6460        6460  WALKING_UPSTAIRS             -1.568669   
15            15          STANDING             -1.573148   
7624        7624           SITTING             -1.570374   
918          918            LAYING             -1.548333   
1932        1932  WALKING_UPSTAIRS             -1.572367   

      angle(Y,gravityMean)  angle(Z,gravityMean)  \
5954             -0.790451             -2.453188   
6460             -0.612620             -2.488338   
15               -0.804714             -2.466698   
7624             -0.807312             -2.432511   
918              -0.139628             -2.380889   
1932             -0.888191             -2.456387   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyAccMean,gravity)  \
5954                             -1.931042                     0.172455   
6460                             -1.751117                     0.218134   
15                               -1.862467                     0.181035   
7624                             -1.836867                     0.150060   
918                              -1.915239                     0.067613   
1932                             -1.761974                     0.033891   

      angle(tBodyGyroJerkMean,gravityMean)  angle(tBodyGyroMean,gravityMean)  \
5954                             -0.336884                          6.739432   
6460                              0.280294                          7.953403   
15                               -0.475323                          4.723190   
7624                             -0.362867                          5.352938   
918                              -0.887342                          6.007361   
1932                             -0.318793                          8.175504   

      energy-mean()  fBodyAcc-bandsEnergy()-1,16  \
5954     -11.961822                     2.662012   
6460      -0.679894                     2.512880   
15        -5.669584                     3.624072   
7624      -4.480885                     3.624099   
918      -11.905892                     3.621984   
1932      -5.075754                     2.051232   

      fBodyAcc-bandsEnergy()-1,16.1  fBodyAcc-bandsEnergy()-1,16.2  \
5954                       5.717897                       2.711735   
6460                       5.740178                       2.803435   
15                         6.158806                       2.495663   
7624                       6.159186                       2.495153   
918                        6.158630                       2.495276   
1932                       5.541644                       2.875009   

      fBodyAcc-bandsEnergy()-1,24  fBodyAcc-bandsEnergy()-1,24.1  \
5954                     5.535808                       1.880821   
6460                     5.484726                       2.102368   
15                       5.777385                       3.315121   
7624                     5.777394                       3.316270   
918                      5.776843                       3.314383   
1932                     5.365202                       1.534785   

      fBodyAcc-bandsEnergy()-1,24.2  fBodyAcc-bandsEnergy()-1,8  \
5954                      -1.354953                   -1.351167   
6460                      -1.355854                   -1.416568   
15                        -1.352863                   -1.537945   
7624                      -1.352859                   -1.537948   
918                       -1.352860                   -1.537483   
1932                      -1.356338                   -1.236738   

      fBodyAcc-bandsEnergy()-1,8.1  fBodyAcc-bandsEnergy()-1,8.2  \
5954                      2.584836                      8.198736   
6460                      2.527046                      8.046001   
15                        2.845798                      8.617922   
7624                      2.846161                      8.

W trakcie analizy dochodzimy do kilku wniosków, mianowicie:

In [22]:
pd.unique(df['Activity'])

array(['STANDING', 'WALKING_UPSTAIRS', 'LAYING', 'WALKING', 'SITTING',
       'WALKING_DOWNSTAIRS'], dtype=object)

1. Mamy 6 aktywności, które nasz model ma rozpoznawać,

In [23]:
df.subject.nunique()

30

2. W badaniu brało udział 30 uczestników,

3. Do trenowania modelu klasyfikacji definitywnie nie potrzebujemy wszystkich cech. Skupiamy się na tych, które zawierają średnią i odchylenie standardowe.

## Oczyszczanie danych

Zanim zaczniemy czyścić zbiór, wyciągamy z niego dane o uczestnikach i aktywnościach, to przyda nam się potem.

In [24]:
subject = df['subject']
activity = df['Activity']

Na początku wybieramy wszystkie kolumny zawierające średnią lub odchylenie standardowe wszystkich z cech.

In [25]:
mean_cols = [col for col in df.columns if 'mean' in col or 'Mean' in col]
std_cols = [col for col in df.columns if 'std' in col or 'Std' in col]

df = pd.concat([df[std_cols], df[mean_cols]], axis = 1)
df.shape

(11086, 87)

In [26]:
573-87, (573-87)/573

(486, 0.8481675392670157)

W ten sposób udało nam się pozbyć 486 cech, co stanowi prawie 85% całego zbioru.

Aktualnie nasz zbiór wygląda tak:

In [27]:
df.sample(4)

fBodyAcc-std()-X  fBodyAcc-std()-Y  fBodyAcc-std()-Z  \
5615          0.807474         -1.286602          4.201082   
2371         -1.483615         -4.065444          3.825157   
3287         -0.365268         -2.889240          3.808517   
2762          0.755296         -1.283612          4.217478   

      fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  fBodyAccJerk-std()-Z  \
5615             -4.017704              0.279585              2.364808   
2371             -6.183370              2.936569              1.603713   
3287             -4.818924              1.251965              1.997194   
2762             -4.063730              0.303414              2.333802   

      fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
5615          -0.816364                  -3.016496   
2371          -3.228954                  -2.799804   
3287          -2.125550                  -2.951323   
2762          -0.806943                  -3.011841   

      fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
5615                   -6.015518                4.850034           3.313035   
2371                   -6.102991                4.625130           1.252464   
3287                   -6.036779                4.703892           1.503267   
2762                   -6.017739                4.845153           3.302522   

      fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  \
5615           0.812341           4.712813          6.349421   
2371           0.742475           4.056954          5.604046   
3287           0.737002           3.962423          5.961208   
2762           0.807615           4.705391          6.331799   

      tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAccJerk-std()-X  \
5615          3.883784          3.432850              6.318784   
2371          3.671329          2.475073              4.674729   
3287          3.770991          2.570863              5.690405   
2762          3.883619          3.456453              6.284308   

      tBodyAccJerk-std()-Y  tBodyAccJerk-std()-Z  tBodyAccJerkMag-std()  \
5615             -0.534556              3.927610               2.311231   
2371             -3.061789              3.322022               3.391179   
3287             -1.419423              3.651714               2.675180   
2762             -0.557388              3.904492               2.332787   

      tBodyAccMag-std()  tBodyGyro-std()-X  tBodyGyro-std()-Y  \
5615           1.628526           5.355272           5.729763   
2371          -1.909790           3.318010           3.649926   
3287          -0.246466           3.788330           3.900962   
2762           1.623716           5.340579           5.605033   

      tBodyGyro-std()-Z  tBodyGyroJerk-std()-X  tBodyGyroJerk-std()-Y  \
5615           1.798291              -4.162572              -0.591802   
2371           1.097928              -2.991235              -0.463604   
3287           1.113066              -3.869637              -0.545497   
2762           1.789814              -4.134095              -0.587106   

      tBodyGyroJerk-std()-Z  tBodyGyroJerkMag-std()  tBodyGyroMag-std()  \
5615               0.852420               -2.669518            5.827425   
2371               0.854747               -2.236299            3.877963   
3287               0.853237               -2.543795            4.753172   
2762               0.852456               -2.656659            5.780229   

      tGravityAcc-std()-X  tGravityAcc-std()-Y  tGravityAcc-std()-Z  \
5615             3.745802             1.420299             4.030802   
2371             3.587250             1.235240             4.104551   
3287             3.614517             1.322574             4.073769   
2762             3.621391             1.430272             4.146776   

      tGravityAccMag-std()  angle(X,gravityMean)  angle(Y,gravityMean)  \
5615              2.030178             -1.573983             -0.784796   
2371             -2.189194             -1.569917         

Dodatkowo zdecydowaliśmy się pozbyć kolumn, które zawierają słowa kluczowe: 'angle', 'energy' i 'meanFreq'

In [28]:
angle_cols = [col for col in df.columns if 'angle' in col or 'Angle' in col]
energy_cols = [col for col in df.columns if 'energy' in col or 'Energy' in col]
meanFreq_cols = [col for col in df.columns if 'meanFreq' in col]

drop_cols = angle_cols + energy_cols + meanFreq_cols
df = df.drop(drop_cols, axis=1)

In [29]:
df.shape

(11086, 66)

Ostatecznie nasz zbiór ma 66 cech.

Dodajemy kolumny uczestnika i aktywności.

In [30]:
df.insert(0, 'Activity', activity)
df.insert(0, 'Subject', subject)

Pozbywamy się niekompletnych wierszy.

In [31]:
df = df.dropna()

Dane, które użyliśmy do przetestowania modelu

In [32]:
df.shape

(9826, 68)

In [33]:
df.head(10)

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0        30            STANDING          0.756296         -1.586674   
1         9    WALKING_UPSTAIRS         -0.537596         -2.784092   
2        28              LAYING          0.755296         -1.283612   
3        25             WALKING          0.149588         -2.796041   
4        28              LAYING          0.755296         -1.283612   
5        23             SITTING          0.664783         -2.040734   
6        27              LAYING          0.802888         -1.258179   
7        14            STANDING          0.794411         -1.344356   
9         7  WALKING_DOWNSTAIRS         -1.309521         -2.932957   
10       17             WALKING         -0.431861         -2.956632   

    fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0           4.141758             -4.176057              0.585073   
1           3.677794             -5.817531              1.293797   
2           4.217478             -4.063730              0.303414   
3           3.922191             -4.776575              1.763192   
4           4.217478             -4.063730              0.303414   
5           4.042778             -4.141761              0.414488   
6           4.220286             -4.023362              0.263590   
7           4.215034             -4.022720              0.291862   
9           3.863861             -6.394034              1.706218   
10          3.794916             -5.436230              2.108401   

    fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0               2.245282          -1.000675                  -2.979941   
1               1.616815          -2.166122                  -2.906000   
2               2.333802          -0.806943                  -3.011841   
3               1.876670          -1.633627                  -2.920310   
4               2.333802          -0.806943                  -3.011841   
5               2.316583          -1.467383                  -2.994722   
6               2.358539          -0.768377                  -3.016589   
7               2.360033          -0.830130                  -3.015687   
9               1.549443          -3.221167                  -2.809464   
10              1.639118          -1.476475                  -2.884412   

    fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                    -6.033122                4.781842           2.565681   
1                    -6.062388                4.653310           1.148811   
2                    -6.017739                4.845153           3.302522   
3                    -6.048324                4.737144           1.710101   
4                    -6.017739                4.845153           3.302522   
5                    -6.027634                4.785031           3.007326   
6                    -6.015245                4.852861           3.342725   
7                    -6.016443                4.843876           3.254727   
9                    -6.100774                4.680560           1.896524   
10                   -6.080514                4.762033           1.975335   

    fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  tBodyAcc-std()-Y  \
0            0.795967           4.632040          6.325173          3.860434   
1            0.660377           3.794268          5.880477          3.775318   
2            0.807615           4.705391          6.331799          3.883619   
3            0.746790           4.400935          6.119588          3.769096   
4            0.807615           4.705391          6.331799          3.883619   
5            0.796710           4.558747          6.302001          3.832121   
6            0.813807           4.718742          6.347577          3.885898   
7            0.812325           4.695774          6.345157          3.879934   
9            0.726239           4.169236          5.663822          3.759751   
10        

Ostatecznie zdecydowaliśmy się zgrupować wszystkich uczestników badania razem z aktywnościami, w których brali udział

In [15]:
grouped = df.groupby(['Subject', 'Activity'])

Wyciągamy średnią z wszystkich cech, dla każdego uczestnika względem każdej z badanych aktywności.

In [16]:
df = grouped.mean().reset_index()

I tak prezentują się nasze zgrupowane oczyszczone dane, które użyliśmy do trenowania modelu:

In [18]:
df.shape

(180, 68)

In [17]:
df.head(10)

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0        1              LAYING          0.772614         -1.312344   
1        1             SITTING          0.798087         -1.307646   
2        1            STANDING          0.802282         -1.294112   
3        1             WALKING         -0.424408         -3.359302   
4        1  WALKING_DOWNSTAIRS         -0.938943         -3.193383   
5        1    WALKING_UPSTAIRS         -0.358470         -3.207309   
6        2              LAYING          0.775048         -1.254903   
7        2             SITTING          0.784312         -1.341305   
8        2            STANDING          0.789539         -1.324953   
9        2             WALKING         -0.151261         -2.896934   

   fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0          4.192530             -4.032061              0.269285   
1          4.214334             -4.022398              0.260490   
2          4.213188             -4.021302              0.268362   
3          3.828758             -5.753965              2.741275   
4          3.840144             -5.907574              2.474086   
5          3.625047             -5.119551              1.600063   
6          4.219320             -4.029690              0.263934   
7          4.200123             -4.032743              0.281445   
8          4.197566             -4.049089              0.304747   
9          3.921515             -5.403117              2.481886   

   fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0              2.358421          -0.846795                  -3.015795   
1              2.365289          -0.804229                  -3.016613   
2              2.364522          -0.805330                  -3.016604   
3              1.604912          -2.147099                  -2.845857   
4              1.407951          -2.649033                  -2.837071   
5              1.937368          -2.063600                  -2.928284   
6              2.361821          -0.785951                  -3.016474   
7              2.358394          -0.868957                  -3.015060   
8              2.350316          -0.848490                  -3.012764   
9              1.742390          -1.741868                  -2.858549   

   fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                   -6.016373                4.840281           3.258072   
1                   -6.015543                4.839362           3.315200   
2                   -6.015761                4.848332           3.322622   
3                   -6.102188                4.649459           1.876334   
4                   -6.111458                4.659888           1.507965   
5                   -6.056443                4.606615           2.011382   
6                   -6.015844                4.843700           3.327517   
7                   -6.016521                4.841547           3.314211   
8                   -6.018500                4.840633           3.278964   
9                   -6.078999                4.746789           2.091507   

   fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  tBodyAcc-std()-Y  \
0           0.812411           4.702018          6.337974          3.882237   
1           0.812666           4.674744          6.346444          3.882630   
2           0.812209           4.704193          6.347743          3.883326   
3           0.654530           4.139078          5.906645          3.724877   
4           0.678322           3.915508          5.741504          3.731460   
5           0.628777           4.280835          5.962865          3.746696   
6           0.811658           4.683152          6.339253          3.885898   
7           0.810556           4.680929          6.341413          3.879928   
8           0.809204           4.688025          6.342291          3.880793   
9           0.737212           4.254019          5.998372      

## Model

Do implementacji modelu użyliśmy biblioteki TensorFlow 

### Definicja modelu

In [ ]:
# build model: 66 -> 150 -> 80 -> 6
model = tf.kras.models.Sequential([
        keras.layers.Dense(150, input_dim=66, activation="relu"),
        keras.layers.Dense(80, activation="relu"),
        keras.layers.Dense(6, activation="softmax")
])

Jest to prosty model. Używa dwóch warstw ukrytych z funkcjami aktywacji ReLU. 

Na ostatniej warstwie jest użyty softmax. Dzięki niej, wartości otrzymane są jak najbardziej czytelne. Sumują się one w przybliżeniu do 1 oraz wartość największa mówi nam, która z czynności jest prawdopodobnie wykonywana.

Dzięki temu możemy w bardziej skomplikowanych implementacjach użyć na przykład drugiej największej wartości.

### Kompilacja modelu

In [ ]:
model.compile(optimizer="adam", loss="MSE", metrics=['accuracy'])

### Trenowanie

Następnym krokiem było trenowanie. Użyliśmy do tego zgrupowanych oczyszczonych danych (6 aktywności na 30 osób - 180 pomiarów)

In [66]:
model.fit(X, Y, epochs=200)

Epoch 1/200
6/6 [==============================] - 0s 609us/step - loss: 0.1620 - accuracy: 0.2444
Epoch 2/200
6/6 [==============================] - 0s 834us/step - loss: 0.1320 - accuracy: 0.3889
Epoch 3/200
6/6 [==============================] - 0s 500us/step - loss: 0.1183 - accuracy: 0.4167
Epoch 4/200
6/6 [==============================] - 0s 834us/step - loss: 0.1049 - accuracy: 0.6444
Epoch 5/200
6/6 [==============================] - 0s 668us/step - loss: 0.0992 - accuracy: 0.6389
Epoch 6/200
6/6 [==============================] - 0s 832us/step - loss: 0.0966 - accuracy: 0.5833
Epoch 7/200
6/6 [==============================] - 0s 499us/step - loss: 0.0904 - accuracy: 0.6667
Epoch 8/200
6/6 [==============================] - 0s 668us/step - loss: 0.0896 - accuracy: 0.6056
Epoch 9/200
6/6 [==============================] - 0s 667us/step - loss: 0.0990 - accuracy: 0.5944
Epoch 10/200
6/6 [==============================] - 0s 666us/step - loss: 0.0926 - accuracy: 0.6167
Epoch 11/

6/6 [==============================] - 0s 835us/step - loss: 0.0147 - accuracy: 0.9333
Epoch 165/200
6/6 [==============================] - 0s 501us/step - loss: 0.0179 - accuracy: 0.9111
Epoch 166/200
6/6 [==============================] - 0s 832us/step - loss: 0.0155 - accuracy: 0.9333
Epoch 167/200
6/6 [==============================] - 0s 833us/step - loss: 0.0160 - accuracy: 0.9389
Epoch 168/200
6/6 [==============================] - 0s 499us/step - loss: 0.0236 - accuracy: 0.8889
Epoch 169/200
6/6 [==============================] - 0s 667us/step - loss: 0.0193 - accuracy: 0.9167
Epoch 170/200
6/6 [==============================] - 0s 1ms/step - loss: 0.0168 - accuracy: 0.9389
Epoch 171/200
6/6 [==============================] - 0s 1ms/step - loss: 0.0149 - accuracy: 0.9500
Epoch 172/200
6/6 [==============================] - 0s 834us/step - loss: 0.0149 - accuracy: 0.9389
Epoch 173/200
6/6 [==============================] - 0s 501us/step - loss: 0.0131 - accuracy: 0.9500
Epoch 17

### Dokładność

Po przetrenowaniu modelu na wcześniejszych danych, sprawdziliśmy dokładność modelu

In [25]:
model.evaluate(X_2, Y_2, verbose=1)


 model evaluation:
308/308 [==============================] - 0s 404us/step - loss: 0.0198 - accuracy: 0.9232


[0.019759194925427437, 0.9231630563735962]

Osiągnęliśmy dokładność rzędu 92% na oczyszczonych danych (wyeliminowanie wierszy z NaNami (9826 pomiarów))

Gotowy model zapisaliśmy do cp.ckpt